urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/util/HexDumpOutputStream.java

```diff
@@ -0,0 +1,307 @@
+	throws IOException {
+
+	if( this.isClosed() )
+	    throw new IOException( "Cannot write to a closed HexDumpOutputStream." );
+
+	this.buffer.write( (byte)(b & 0xFF) );
+
+	// Check if the buffer is full (then flush).
+	if( this.buffer.isFull() )
+	    this.flush();
+	    
+    }
+    //--- END ----------------------- Writer implementation --------------------
+
+    private void writeOffset() 
+	throws IOException {
+
+	this.coreWriter.write( "0x" );
+	String addr =
+	    Integer.toString( (int)( (this.writeOffset>>28) & 0xF ), 16 ) +
+	    Integer.toString( (int)( (this.writeOffset>>24) & 0xF ), 16 ) +
+	    Integer.toString( (int)( (this.writeOffset>>20) & 0xF ), 16 ) +
+	    Integer.toString( (int)( (this.writeOffset>>16) & 0xF ), 16 ) +
+	    Integer.toString( (int)( (this.writeOffset>>12) & 0xF ), 16 ) +
+	    Integer.toString( (int)( (this.writeOffset>> 8) & 0xF ), 16 ) +
+	    Integer.toString( (int)( (this.writeOffset>> 4) & 0xF ), 16 ) +
+	    Integer.toString( (int)( (this.writeOffset    ) & 0xF ), 16 );
+
+	
+	// Fill the address prefix with zeros.
+	for( int i = addr.length(); i < 8; i++ )
+	    this.coreWriter.write( "0" );
+
+	// Print the actual address part and a delimiter
+	this.coreWriter.write( addr );
+
+    }
+
+    private int writeHexView() 
+	throws IOException {
+
+	// Transfer all bytes from the buffer to the underlying writer.
+	byte b;
+	int i = 0;
+	for( int block = 0; block < this.blockSizes.length; block++ ) {
+
+	    int blockSize = this.blockSizes[block];
+	    //int inBlock = 0;
+
+	    //while( inBlock < blockSize ) {
+	    for( int inBlock = 0; inBlock < blockSize; inBlock++ ) {
+
+		if( i < this.buffer.length() ) {
+		    b = this.buffer.peek( i );
+		    this.coreWriter.write( Integer.toString( (b>>4) & 0xF, 16 ) );
+		    this.coreWriter.write( Integer.toString(  b     & 0xF, 16 ) );	
+		    
+		    i++;
+		} else {
+		    
+		    // Write fill spaces for the case we reached end-of-file (EOI)
+		    this.coreWriter.write( "  " ); // Each byte has two (!) hex chars
+
+		}
+		
+		//inBlock++;
+	    }
+	    
+	    // Print a block delimiter
+	    this.coreWriter.write( " " ); // Space
+
+	}
+
+
+	return i;
+    }
+
+    private void printCleanView() 
+	throws IOException {
+
+	byte b;
+	char c;
+	while( !this.buffer.isEmpty() ) {
+
+	    b = this.buffer.read();
+	    c = (char)b;
+	    // Only write printable (non control) characters
+	    if( ( c > 32  
+		
+		  // At position 126 is the DEL control char
+		  && c < 126 )
+		
+		||
+		(c > 126 && c < 178) ) {
+		
+		// Print a placeholder
+		this.coreWriter.write( c );
+
+	    } else {
+
+		// Print a placeholder
+		this.coreWriter.write( "." );
+
+	    }
+	}
+    }
+
+    public static void main( String[] argv ) {
+
+	if( argv.length == 0 ) {
+
+	    System.err.println( "Specify an input file." );
+	    System.exit( 1 );
+
+	}
+
+
+	try {
+
+
+	    java.io.InputStream in = new java.io.FileInputStream( new java.io.File(argv[0]) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at line 287. The resource is closed at line 293. There are other execution paths that do not contain closure statements, for example, when InputStream.read throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).